File: src/main/java/dominio/AbstractLocalRepository.java
package dominio;

import excepciones.EntidadExistenteError;

import javax.persistence.*;
import java.util.HashSet;
import java.util.List;
import java.util.Set;
import java.util.stream.Collectors;

@MappedSuperclass
public abstract class AbstractLocalRepository<T>{
	
	@Id 
	@GeneratedValue
	private Long id;
	
	@OneToMany(fetch = FetchType.EAGER, cascade = CascadeType.ALL, orphanRemoval = true)
	@JoinColumn(name="usuario_id")
	protected Set<T> elementos = new HashSet<>();
	
	public List<T> obtenerTodos() {
		return elementos.stream().collect(Collectors.toList());
	}
	
	public void agregar(T elemento){
		if (existe(elemento)){
			throw new EntidadExistenteError(mensajeEntidadExistenteError(elemento));
		}
		elementos.add(elemento);
	}
	
	protected boolean existe(T elemento){
		return this.obtenerTodos().contains(elemento);//Contains chequea con equals
	}
	
	protected abstract String mensajeEntidadExistenteError(T elemento);
}

File: src/main/java/dondeInvierto/dominio/repo/AbstractLocalRepository.kt
package dondeInvierto.dominio.repo

import excepciones.EntidadExistenteError
import java.util.*
import javax.persistence.*

@MappedSuperclass
abstract class AbstractLocalRepository<T> {

    @Id
    @GeneratedValue
    private val id: Long? = null

    @OneToMany(fetch = FetchType.EAGER, cascade = [CascadeType.ALL], orphanRemoval = true)
    @JoinColumn(name = "usuario_id")
    protected var elementos: MutableSet<T> = HashSet()

    fun obtenerTodos() = elementos.toList()

    fun agregar(elemento: T) {
        if (existe(elemento)) {
            throw EntidadExistenteError(mensajeEntidadExistenteError(elemento))
        }
        elementos.add(elemento)
    }

    protected fun existe(elemento: T) = obtenerTodos().contains(elemento)

    protected abstract fun mensajeEntidadExistenteError(elemento: T): String
}
